File: src/main/java/ru/evotor/framework/inventory/provider/InventoryContract.java
package ru.evotor.framework.inventory.provider;

import android.net.Uri;

import ru.evotor.framework.core.DoNotUseThis;
import ru.evotor.framework.provider.IdentifiedEntityColumns;

@DoNotUseThis()
public final class InventoryContract {
    private InventoryContract() {
    }

    public static final String AUTHORITY = "ru.evotor.framework.inventory";

    public static final Uri AUTHORITY_URI = Uri.parse("content://" + AUTHORITY);

    public interface ProductColumns extends IdentifiedEntityColumns {
        String VARIATION_ID = "VARIATION_ID";
        String GROUP_UUID = "GROUP_UUID";
        String NAME = "NAME";
        String CODE = "CODE";
        String VENDOR_CODE = "VENDOR_CODE";
        String PRICE = "PRICE";
        String VAT_RATE = "VAT_RATE";
        String QUANTITY_EXACT_VALUE = "QUANTITY_EXACT_VALUE";
        String QUANTITY_SCALE = "QUANTITY_SCALE";
        String UNIT_OF_MEASUREMENT_VARIATION_ID = "UNIT_OF_MEASUREMENT_VARIATION_ID";
        String UNIT_OF_MEASUREMENT_TYPE = "UNIT_OF_MEASUREMENT_TYPE";
        String UNIT_OF_MEASUREMENT_NAME = "UNIT_OF_MEASUREMENT_NAME";
        String DESCRIPTION = "DESCRIPTION";
    }

    public interface ProductExtensionColumns {
        String PRODUCT_UUID = "PRODUCT_UUID";
    }

    public interface AlcoholProductColumns extends ProductExtensionColumns {
        String FSRAR_PRODUCT_KIND_CODE = "FSRAR_PRODUCT_KIND_CODE";
        String TARE_VOLUME_EXACT_VALUE = "TARE_VOLUME_EXACT_VALUE";
        String TARE_VOLUME_SCALE = "TARE_VOLUME_SCALE";
        String TARE_VOLUME_UNIT_OF_MEASUREMENT_VARIATION_ID = "TARE_VOLUME_UNIT_OF_MEASUREMENT_VARIATION_ID";
        String TARE_VOLUME_UNIT_OF_MEASUREMENT_NAME = "TARE_VOLUME_UNIT_OF_MEASUREMENT_NAME";
        String ALCOHOL_PERCENTAGE = "ALCOHOL_PERCENTAGE";
    }

    public static final class Product implements ProductColumns, AlcoholProductColumns {
        private Product() {

        }

        public static final Uri CONTENT_URI = Uri.withAppendedPath(AUTHORITY_URI, "products");

        public static final String CONTENT_TYPE = "vnd.android.cursor.dir/vnd.ru.evotor.framework.inventory.product.Product";

        public static final Uri UNCLASSIFIED_PRODUCTS_URI = Uri.withAppendedPath(CONTENT_URI, "unclassified_products");
        public static final Uri UNCLASSIFIED_SERVICES_URI = Uri.withAppendedPath(CONTENT_URI, "unclassified_services");
        public static final Uri WEAK_ALCOHOL_URI = Uri.withAppendedPath(CONTENT_URI, "weak_alcohol");
        public static final Uri STRONG_ALCOHOL_URI = Uri.withAppendedPath(CONTENT_URI, "strong_alcohol");
        public static final Uri TOBACCO_URI = Uri.withAppendedPath(CONTENT_URI, "tobacco");

        public static final String VARIATION_ID_UNCLASSIFIED_PRODUCT = "NORMAL";
        public static final String VARIATION_ID_UNCLASSIFIED_SERVICE = "SERVICE";
        public static final String VARIATION_ID_WEAK_ALCOHOL = "ALCOHOL_NOT_MARKED";
        public static final String VARIATION_ID_STRONG_ALCOHOL = "ALCOHOL_MARKED";
        public static final String VARIATION_ID_TOBACCO = "TOBACCO_MARKED";
    }

    public static final class ProductExtension implements AlcoholProductColumns {
        private ProductExtension() {

        }

        public static final Uri CONTENT_URI = Uri.withAppendedPath(AUTHORITY_URI, "product_extensions");

        public static final String CONTENT_TYPE = "vnd.android.cursor.dir/vnd.ru.evotor.framework.inventory.product.extension.ProductExtension";

        public static final Uri ALCOHOL_PRODUCTS_URI = Uri.withAppendedPath(CONTENT_URI, "alcohol_products");
    }

    public interface ProductGroupColumns extends IdentifiedEntityColumns {
        String PARENT_GROUP_UUID = "PARENT_GROUP_UUID";
        String NAME = "NAME";
    }

    public static final class ProductGroup implements ProductGroupColumns {
        private ProductGroup() {

        }

        public static final Uri CONTENT_URI = Uri.withAppendedPath(AUTHORITY_URI, "product_groups");

        public static final String CONTENT_TYPE = "vnd.android.cursor.dir/vnd.ru.evotor.framework.inventory.ProductGroup";
    }

    public interface BarcodeColumns {
        String PRODUCT_UUID = "COMMODITY_UUID";
        String VALUE = "BARCODE";
    }

    public static final class Barcode implements BarcodeColumns {
        private Barcode() {

        }

        public static final Uri CONTENT_URI = Uri.withAppendedPath(AUTHORITY_URI, "barcodes");

        public static final String CONTENT_TYPE = "vnd.android.cursor.dir/vnd.ru.evotor.framework.inventory.product.Barcode";

        public static Uri getParticularBarcodeUri(String barcode) {
            return Uri.withAppendedPath(CONTENT_URI, barcode);
        }

        public static Uri getProductsByBarcodeUri(String barcode) {
            return Uri.withAppendedPath(getParticularBarcodeUri(barcode), "products");
        }

        public static Uri getPositionsByBarcodeUri(String barcode) {
            return Uri.withAppendedPath(getParticularBarcodeUri(barcode), "positions");
        }
    }
}

File: src/main/java/ru/evotor/framework/inventory/provider/InventoryContract.kt
package ru.evotor.framework.inventory.provider

import android.net.Uri

object InventoryContract {
    const val AUTHORITY = "ru.evotor.framework.inventory"

    val BASE_URI: Uri = Uri.parse("content://$AUTHORITY")

    fun getBarcodeUri(barcode: String): Uri = Uri.withAppendedPath(BASE_URI, barcode)

    const val PATH_PRODUCTS = "products"

    const val PATH_POSITIONS = "positions"

    val URI_PRODUCTS: Uri = Uri.withAppendedPath(BASE_URI, PATH_PRODUCTS)

    const val PATH_UNCLASSIFIED_PRODUCTS = "unclassified_products"
    const val PATH_UNCLASSIFIED_PAYABLE_SERVICES = "unclassified_payable_services"
    const val PATH_WEAK_ALCOHOL = "weak_alcohol"
    const val PATH_STRONG_ALCOHOL = "strong_alcohol"
    const val PATH_TOBACCO = "tobacco"

    const val PATH_PRODUCT_EXTENSIONS = "product_extensions"

    val URI_PRODUCT_EXTENSIONS: Uri = Uri.withAppendedPath(BASE_URI, PATH_PRODUCT_EXTENSIONS)

    const val PATH_ALCOHOL_PRODUCTS = "alcohol_products"

    const val PATH_PRODUCT_GROUPS = "product_groups"

    object ProductColumns {
        const val VARIATION_ID = "VARIATION_ID"

        const val VARIATION_ID_UNCLASSIFIED_PRODUCT = "NORMAL"
        const val VARIATION_ID_UNCLASSIFIED_PAYABLE_SERVICE = "SERVICE"
        const val VARIATION_ID_WEAK_ALCOHOL = "ALCOHOL_NOT_MARKED"
        const val VARIATION_ID_STRONG_ALCOHOL = "ALCOHOL_MARKED"
        const val VARIATION_ID_TOBACCO = "TOBACCO"

        const val UUID = "UUID"
        const val GROUP_UUID = "GROUP_UUID"
        const val NAME = "NAME"
        const val CODE = "CODE"
        const val VENDOR_CODE = "VENDOR_CODE"
        const val BARCODES = "BARCODES"
        const val PURCHASE_PRICE = "PURCHASE_PRICE"
        const val SELLING_PRICE = "SELLING_PRICE"
        const val VAT_RATE = "VAT_RATE"
        const val QUANTITY = "QUANTITY"
        const val DESCRIPTION = "DESCRIPTION"
        const val ALLOWED_TO_SELL = "ALLOWED_TO_SELL"
    }

    object UnitOfMeasurementColumns {
        const val VARIATION_ID = "UNIT_OF_MEASUREMENT_VARIATION_ID"

        const val VARIATION_ID_CUSTOM = 0
        const val VARIATION_ID_CONVENTIONAL_UNIT = 1
        const val VARIATION_ID_PIECE = 2
        const val VARIATION_ID_PACKAGING = 3
        const val VARIATION_ID_KIT = 4
        const val VARIATION_ID_KILOGRAM = 5
        const val VARIATION_ID_METER = 6
        const val VARIATION_ID_SQUARE_METER = 7
        const val VARIATION_ID_CUBIC_METER = 8
        const val VARIATION_ID_LITER = 9

        const val TYPE = "UNIT_OF_MEASUREMENT_TYPE"
        const val NAME = "UNIT_OF_MEASUREMENT_NAME"
        const val PRECISION = "UNIT_OF_MEASUREMENT_PRECISION"
    }

    object AlcoholProductColumns {
        const val PRODUCT_UUID = "PRODUCT_UUID"
        const val FSAR_PRODUCT_KIND_CODE = "FSAR_PRODUCT_KIND_CODE"
        const val TARE_VOLUME = "TARE_VOLUME"
        const val ALCOHOL_PERCENTAGE = "ALCOHOL_PERCENTAGE"
    }

    object ProductGroupColumns {
        const val UUID = "UUID"
        const val PARENT_GROUP_UUID = "PARENT_GROUP_UUID"
        const val NAME = "NAME"
    }
}